lib/events/export/date_exporter.go

```diff
@@ -0,0 +1,461 @@
+		if ok := e.startProcessingChunk(ctx, chunk, cursor); !ok {
+			return
+		}
+	}
+
+	for {
+		n, err := e.fetchAndProcessChunks(ctx)
+		if err != nil {
+			if ctx.Err() != nil {
+				return
+			}
+
+			if e.mainLogLimiter.Allow() {
+				e.log.WarnContext(ctx, "fetch and process of event chunks failed", "error", err)
+			}
+
+			retry.Inc()
+			select {
+			case <-retry.After():
+			case <-ctx.Done():
+				return
+			}
+			continue
+		}
+
+		retry.Reset()
+
+		// if no new chunks were processed, the exporter is considered idle
+		idle := n == 0
+		e.withLock(func() {
+			e.idle = idle
+		})
+		if idle && e.cfg.OnIdle != nil {
+			e.cfg.OnIdle(ctx)
+		}
+
+		// log first success, and periodically log subsequent non-idle cycles
+		if (n > 0 && e.mainLogLimiter.Allow()) || !firstFullCycleCompleted {
+			e.log.InfoContext(ctx, "successful fetch and process of event chunks", "chunks", n, "idle", idle)
+		}
+
+		firstFullCycleCompleted = true
+
+		select {
+		case <-poll.Next():
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// waitForInflightChunks blocks until all inflight chunks have been processed by acquiring all
+// semaphore tokens and then releasing them. note that this operation does not accept a context,
+// which is necessary in order to ensure that Done actually waits for all background processing
+// to halt.
+func (e *DateExporter) waitForInflightChunks() {
+	// acquire all semaphore tokens to block until all inflight chunks have been processed
+	for i := 0; i < e.cfg.Concurrency; i++ {
+		e.sem <- struct{}{}
+	}
+
+	// release all semaphore tokens
+	for i := 0; i < e.cfg.Concurrency; i++ {
+		<-e.sem
+	}
+}
+
+// fetchAndProcessChunks fetches and processes all chunks for the current date. if the function returns
+// without error, all chunks have been successfully processed. note that all *currently known* chunks being
+// processed does not necessarily imply that all events for a given date have been processed since there may
+// be event replication delays, and/or ongoing activity if the current date is being polled.
+func (e *DateExporter) fetchAndProcessChunks(ctx context.Context) (int, error) {
+	// wait for inflight chunks before returning. in theory it would be fine (and possibly more performant)
+	// to return immediately, but doing so makes it difficult to reason about when the exporter has fully exited
+	// and/or how many complete export cycles have been performed.
+	defer e.waitForInflightChunks()
+
+	chunks := e.cfg.Client.GetEventExportChunks(ctx, &auditlogpb.GetEventExportChunksRequest{
+		Date: timestamppb.New(e.cfg.Date),
+	})
+
+	var newChunks int
+
+Chunks:
+	for chunks.Next() {
+		// known chunks should be skipped
+		var skip bool
+		e.withLock(func() {
+			if _, ok := e.chunks[chunks.Item().Chunk]; ok {
+				skip = true
+				return
+			}
+
+			// so long as there is at least one undiscovered chunk, the exporter is not considered idle.
+			e.idle = false
+		})
+
+		if skip {
+			continue Chunks
```

Comment:
Is the `Chunks` label necessary here? It looks like a normal `continue` would behave the same